File: packages/sdk-node/src/webhooks.ts
import type { IncomingMessage, ServerResponse } from 'http';

// ─────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────

/** Attachment from an operator message */
export interface OperatorAttachment {
  filename: string;
  mimeType: string;
  size: number;
  data: Buffer;
  bridgeFileId?: string;
}

/** Callback when operator sends a message from a bridge */
export type OperatorMessageCallback = (
  sessionId: string,
  content: string,
  operatorName: string,
  sourceBridge: 'telegram' | 'discord' | 'slack',
  attachments: OperatorAttachment[],
  replyToBridgeMessageId?: number | null,
  bridgeMessageId?: number | string | null
) => void | Promise<void>;

export type OperatorMessageEditCallback = (
  sessionId: string,
  bridgeMessageId: number | string,
  content: string,
  sourceBridge: 'telegram' | 'discord' | 'slack',
  editedAt?: string
) => void | Promise<void>;

export type OperatorMessageDeleteCallback = (
  sessionId: string,
  bridgeMessageId: number | string,
  sourceBridge: 'telegram' | 'discord' | 'slack',
  deletedAt?: string
) => void | Promise<void>;

/** Webhook handler configuration */
export interface WebhookConfig {
  telegramBotToken?: string;
  slackBotToken?: string;
  discordBotToken?: string;
  allowedBotIds?: string[];
  onOperatorMessage: OperatorMessageCallback;
  onOperatorMessageEdit?: OperatorMessageEditCallback;
  onOperatorMessageDelete?: OperatorMessageDeleteCallback;
}

// ─────────────────────────────────────────────────────────────────
// Telegram Types
// ─────────────────────────────────────────────────────────────────

interface TelegramUpdate {
  update_id: number;
  message?: TelegramMessage;
  edited_message?: TelegramMessage;
}

interface TelegramMessage {
  message_id: number;
  message_thread_id?: number;
  chat: { id: number };
  from?: { id: number; first_name: string; username?: string };
  text?: string;
  caption?: string;
  photo?: TelegramPhotoSize[];
  document?: TelegramDocument;
  audio?: TelegramAudio;
  video?: TelegramVideo;
  voice?: TelegramVoice;
  reply_to_message?: { message_id: number };
  date: number;
  edit_date?: number;
}

interface TelegramPhotoSize {
  file_id: string;
  file_unique_id: string;
  width: number;
  height: number;
  file_size?: number;
}

interface TelegramDocument {
  file_id: string;
  file_name?: string;
  mime_type?: string;
  file_size?: number;
}

interface TelegramAudio {
  file_id: string;
  file_name?: string;
  mime_type?: string;
  file_size?: number;
}

interface TelegramVideo {
  file_id: string;
  file_name?: string;
  mime_type?: string;
  file_size?: number;
}

interface TelegramVoice {
  file_id: string;
  mime_type?: string;
  file_size?: number;
}

interface ParsedMedia {
  fileId: string;
  filename: string;
  mimeType: string;
  size: number;
}

// ─────────────────────────────────────────────────────────────────
// Slack Types
// ─────────────────────────────────────────────────────────────────

interface SlackEventPayload {
  type: string;
  token?: string;
  challenge?: string;
  team_id?: string;
  event?: SlackEvent;
}

interface SlackEvent {
  type: string;
  channel?: string;
  user?: string;
  text?: string;
  ts?: string;
  thread_ts?: string;
  bot_id?: string;
  subtype?: string;
  files?: SlackFile[];
  message?: {
    text?: string;
    user?: string;
    bot_id?: string;
    ts?: string;
    thread_ts?: string;
    files?: SlackFile[];
  };
  previous_message?: {
    text?: string;
    user?: string;
    bot_id?: string;
    ts?: string;
    thread_ts?: string;
  };
  deleted_ts?: string;
}

interface SlackFile {
  id: string;
  name: string;
  mimetype: string;
  size: number;
  url_private: string;
  url_private_download?: string;
}

// ─────────────────────────────────────────────────────────────────
// Discord Types
// ─────────────────────────────────────────────────────────────────

interface DiscordInteraction {
  type: number;
  id: string;
  application_id: string;
  token: string;
  channel_id?: string;
  guild_id?: string;
  member?: { user?: DiscordUser };
  user?: DiscordUser;
  data?: DiscordInteractionData;
}

interface DiscordUser {
  id: string;
  username: string;
}

interface DiscordInteractionData {
  name?: string;
  custom_id?: string;
  options?: Array<{ name: string; value: string }>;
}

const DISCORD_INTERACTION_PING = 1;
const DISCORD_INTERACTION_APPLICATION_COMMAND = 2;
const DISCORD_RESPONSE_PONG = 1;
const DISCORD_RESPONSE_CHANNEL_MESSAGE = 4;

// ─────────────────────────────────────────────────────────────────
// Webhook Handler
// ─────────────────────────────────────────────────────────────────

export class WebhookHandler {
  private config: WebhookConfig;

  constructor(config: WebhookConfig) {
    this.config = config;
  }

  /**
   * Create an Express/Connect middleware for handling Telegram webhooks
   */
  handleTelegramWebhook(): (req: IncomingMessage & { body?: unknown }, res: ServerResponse) => Promise<void> {
    return async (req, res) => {
      if (!this.config.telegramBotToken) {
        res.statusCode = 404;
        res.end(JSON.stringify({ error: 'Telegram not configured' }));
        return;
      }

      try {
        const body = await this.parseBody(req);
        const update = body as TelegramUpdate;

        if (update.edited_message) {
          const msg = update.edited_message;

          if (msg.text?.startsWith('/')) {
            this.writeOK(res);
            return;
          }

          const text = msg.text ?? msg.caption ?? '';
          if (!text) {
            this.writeOK(res);
            return;
          }

          const topicId = msg.message_thread_id;
          if (!topicId) {
            this.writeOK(res);
            return;
          }

          if (this.config.onOperatorMessageEdit) {
            const editedAt = msg.edit_date
              ? new Date(msg.edit_date * 1000).toISOString()
              : new Date().toISOString();
            await this.config.onOperatorMessageEdit(
              String(topicId),
              msg.message_id,
              text,
              'telegram',
              editedAt
            );
          }

          this.writeOK(res);
          return;
        }

        if (update.message) {
          const msg = update.message;

          // Skip commands
          if (msg.text?.startsWith('/')) {
            this.writeOK(res);
            return;
          }

          // Get text content
          const text = msg.text ?? msg.caption ?? '';

          // Parse media
          let media: ParsedMedia | null = null;
          if (msg.photo && msg.photo.length > 0) {
            const largest = msg.photo[msg.photo.length - 1];
            media = {
              fileId: largest.file_id,
              filename: `photo_${Date.now()}.jpg`,
              mimeType: 'image/jpeg',
              size: largest.file_size ?? 0,
            };
          } else if (msg.document) {
            media = {
              fileId: msg.document.file_id,
              filename: msg.document.file_name ?? `document_${Date.now()}`,
              mimeType: msg.document.mime_type ?? 'application/octet-stream',
              size: msg.document.file_size ?? 0,
            };
          } else if (msg.audio) {
            media = {
              fileId: msg.audio.file_id,
              filename: msg.audio.file_name ?? `audio_${Date.now()}.mp3`,
              mimeType: msg.audio.mime_type ?? 'audio/mpeg',
              size: msg.audio.file_size ?? 0,
            };
          } else if (msg.video) {
            media = {
              fileId: msg.video.file_id,
              filename: msg.video.file_name ?? `video_${Date.now()}.mp4`,
              mimeType: msg.video.mime_type ?? 'video/mp4',
              size: msg.video.file_size ?? 0,
            };
          } else if (msg.voice) {
            media = {
              fileId: msg.voice.file_id,
              filename: `voice_${Date.now()}.ogg`,
              mimeType: msg.voice.mime_type ?? 'audio/ogg',
              size: msg.voice.file_size ?? 0,
            };
          }

          // Skip if no content
          if (!text && !media) {
            this.writeOK(res);
            return;
          }

          // Get topic ID (session identifier)
          const topicId = msg.message_thread_id;
          if (!topicId) {
            this.writeOK(res);
            return;
          }

          // Get operator name
          const operatorName = msg.from?.first_name ?? 'Operator';

          // Get reply_to_message ID if present (for visual reply linking)
          const replyToBridgeMessageId = msg.reply_to_message?.message_id ?? null;

          // Download media if present
          const attachments: OperatorAttachment[] = [];
          if (media) {
            const data = await this.downloadTelegramFile(media.fileId);
            if (data) {
              attachments.push({
                filename: media.filename,
                mimeType: media.mimeType,
                size: media.size,
                data,
                bridgeFileId: media.fileId,
              });
            }
          }

          // Call callback
          await this.config.onOperatorMessage(
            String(topicId),
            text,
            operatorName,
            'telegram',
            attachments,
            replyToBridgeMessageId,
            msg.message_id
          );
        }

        this.writeOK(res);
      } catch (error) {
        console.error('[WebhookHandler] Telegram error:', error);
        res.statusCode = 500;
        res.end(JSON.stringify({ error: 'Internal server error' }));
      }
    };
  }

  /**
   * Create an Express/Connect middleware for handling Slack webhooks
   */
  handleSlackWebhook(): (req: IncomingMessage & { body?: unknown }, res: ServerResponse) => Promise<void> {
    return async (req, res) => {
      if (!this.config.slackBotToken) {
        res.statusCode = 404;
        res.end(JSON.stringify({ error: 'Slack not configured' }));
        return;
      }

      try {
        const body = await this.parseBody(req);
        const payload = body as SlackEventPayload;

        // Handle URL verification challenge
        if (payload.type === 'url_verification' && payload.challenge) {
          res.setHeader('Content-Type', 'application/json');
          res.end(JSON.stringify({ challenge: payload.challenge }));
          return;
        }

        // Handle event callbacks
        if (payload.type === 'event_callback' && payload.event) {
          const event = payload.event;

          const isAllowedBot = (botId?: string) =>
            !!botId && (this.config.allowedBotIds?.includes(botId) ?? false);

          if (event.subtype === 'message_changed') {
            if (!this.config.onOperatorMessageEdit) {
              this.writeOK(res);
              return;
            }

            const botId = event.message?.bot_id ?? event.previous_message?.bot_id ?? event.bot_id;
            if (botId && !isAllowedBot(botId)) {
              this.writeOK(res);
              return;
            }

            const threadTs = event.message?.thread_ts ?? event.previous_message?.thread_ts;
            const messageTs = event.message?.ts ?? event.previous_message?.ts;
            const text = event.message?.text ?? '';

            if (threadTs && messageTs) {
              await this.config.onOperatorMessageEdit(
                threadTs,
                messageTs,
                text,
                'slack',
                new Date().toISOString()
              );
            }

            this.writeOK(res);
            return;
          }

          if (event.subtype === 'message_deleted') {
            if (!this.config.onOperatorMessageDelete) {
              this.writeOK(res);
              return;
            }

            const botId = event.previous_message?.bot_id ?? event.bot_id;
            if (botId && !isAllowedBot(botId)) {
              this.writeOK(res);
              return;
            }

            const threadTs = event.previous_message?.thread_ts;
            const messageTs = event.deleted_ts ?? event.previous_message?.ts;

            if (threadTs && messageTs) {
              await this.config.onOperatorMessageDelete(
                threadTs,
                messageTs,
                'slack',
                new Date().toISOString()
              );
            }

            this.writeOK(res);
            return;
          }

          const hasContent =
            event.type === 'message' &&
            event.thread_ts &&
            (!event.bot_id || isAllowedBot(event.bot_id)) &&
            !event.subtype;
          const hasFiles = event.files && event.files.length > 0;

          if (hasContent && (event.text || hasFiles)) {
            const threadTs = event.thread_ts;
            const text = event.text ?? '';

            // Download files if present
            const attachments: OperatorAttachment[] = [];
            if (hasFiles && event.files) {
              for (const file of event.files) {
                const data = await this.downloadSlackFile(file);
                if (data) {
                  attachments.push({
                    filename: file.name,
                    mimeType: file.mimetype,
                    size: file.size,
                    data,
                    bridgeFileId: file.id,
                  });
                }
              }
            }

            // Get operator name
            let operatorName = 'Operator';
            if (event.user) {
              const name = await this.getSlackUserName(event.user);
              if (name) operatorName = name;
            }

            // Call callback (Slack reply support TODO)
            await this.config.onOperatorMessage(
              threadTs!,
              text,
              operatorName,
              'slack',
              attachments,
              null,
              event.ts ?? null
            );
          }
        }

        this.writeOK(res);
      } catch (error) {
        console.error('[WebhookHandler] Slack error:', error);
        res.statusCode = 500;
        res.end(JSON.stringify({ error: 'Internal server error' }));
      }
    };
  }

  /**
   * Create an Express/Connect middleware for handling Discord webhooks
   */
  handleDiscordWebhook(): (req: IncomingMessage & { body?: unknown }, res: ServerResponse) => Promise<void> {
    return async (req, res) => {
      try {
        const body = await this.parseBody(req);
        const interaction = body as DiscordInteraction;

        // Handle PING (verification)
        if (interaction.type === DISCORD_INTERACTION_PING) {
          res.setHeader('Content-Type', 'application/json');
          res.end(JSON.stringify({ type: DISCORD_RESPONSE_PONG }));
          return;
        }

        // Handle Application Commands (slash commands)
        if (
          interaction.type === DISCORD_INTERACTION_APPLICATION_COMMAND &&
          interaction.data
        ) {
          if (interaction.data.name === 'reply') {
            const threadId = interaction.channel_id;
            const content = interaction.data.options?.find(
              (opt) => opt.name === 'message'
            )?.value;

            if (threadId && content) {
              // Get operator name
              const operatorName =
                interaction.member?.user?.username ??
                interaction.user?.username ??
                'Operator';

              // Call callback (Discord reply support TODO)
              await this.config.onOperatorMessage(
                threadId,
                content,
                operatorName,
                'discord',
                [],
                null
              );

              // Respond to interaction
              res.setHeader('Content-Type', 'application/json');
              res.end(
                JSON.stringify({
                  type: DISCORD_RESPONSE_CHANNEL_MESSAGE,
                  data: { content: '✅ Message sent to visitor' },
                })
              );
              return;
            }
          }
        }

        // Default response
        res.setHeader('Content-Type', 'application/json');
        res.end(JSON.stringify({ type: DISCORD_RESPONSE_PONG }));
      } catch (error) {
        console.error('[WebhookHandler] Discord error:', error);
        res.statusCode = 500;
        res.end(JSON.stringify({ error: 'Internal server error' }));
      }
    };
  }

  // ─────────────────────────────────────────────────────────────────
  // Helper Methods
  // ─────────────────────────────────────────────────────────────────

  private async parseBody(req: IncomingMessage & { body?: unknown }): Promise<unknown> {
    if (req.body) return req.body;

    return new Promise((resolve, reject) => {
      let data = '';
      req.on('data', (chunk) => (data += chunk));
      req.on('end', () => {
        try {
          resolve(data ? JSON.parse(data) : {});
        } catch {
          reject(new Error('Invalid JSON'));
        }
      });
      req.on('error', reject);
    });
  }

  private writeOK(res: ServerResponse): void {
    res.setHeader('Content-Type', 'application/json');
    res.end(JSON.stringify({ ok: true }));
  }

  private async downloadTelegramFile(fileId: string): Promise<Buffer | null> {
    try {
      const botToken = this.config.telegramBotToken!;

      // Get file path
      const getFileUrl = `https://api.telegram.org/bot${botToken}/getFile?file_id=${fileId}`;
      const getFileResp = await fetch(getFileUrl);
      const getFileResult = (await getFileResp.json()) as {
        ok: boolean;
        result?: { file_path: string };
      };

      if (!getFileResult.ok || !getFileResult.result?.file_path) {
        return null;
      }

      // Download file
      const downloadUrl = `https://api.telegram.org/file/bot${botToken}/${getFileResult.result.file_path}`;
      const downloadResp = await fetch(downloadUrl);
      const arrayBuffer = await downloadResp.arrayBuffer();
      return Buffer.from(arrayBuffer);
    } catch (error) {
      console.error('[WebhookHandler] Telegram file download error:', error);
      return null;
    }
  }

  private async downloadSlackFile(file: SlackFile): Promise<Buffer | null> {
    try {
      const downloadUrl = file.url_private_download ?? file.url_private;
      const resp = await fetch(downloadUrl, {
        headers: {
          Authorization: `Bearer ${this.config.slackBotToken}`,
        },
      });

      if (!resp.ok) {
        return null;
      }

      const arrayBuffer = await resp.arrayBuffer();
      return Buffer.from(arrayBuffer);
    } catch (error) {
      console.error('[WebhookHandler] Slack file download error:', error);
      return null;
    }
  }

  private async getSlackUserName(userId: string): Promise<string | null> {
    try {
      const url = `https://slack.com/api/users.info?user=${userId}`;
      const resp = await fetch(url, {
        headers: {
          Authorization: `Bearer ${this.config.slackBotToken}`,
        },
      });

      const result = (await resp.json()) as {
        ok: boolean;
        user?: { real_name?: string; name?: string };
      };

      if (!result.ok) {
        return null;
      }

      return result.user?.real_name ?? result.user?.name ?? null;
    } catch {
      return null;
    }
  }
}

File: packages/sdk-ruby/lib/pocketping/webhooks.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

module PocketPing
  # Attachment from an operator message received via webhook
  class OperatorAttachment
    attr_reader :filename, :mime_type, :size, :data, :bridge_file_id

    def initialize(filename:, mime_type:, size:, data:, bridge_file_id: nil)
      @filename = filename
      @mime_type = mime_type
      @size = size
      @data = data
      @bridge_file_id = bridge_file_id
    end

    def to_h
      {
        filename: @filename,
        mimeType: @mime_type,
        size: @size,
        bridgeFileId: @bridge_file_id
      }
    end
  end

  # Configuration for webhook handlers
  class WebhookConfig
    attr_reader :telegram_bot_token, :slack_bot_token, :discord_bot_token
    attr_accessor :on_operator_message, :on_operator_message_with_ids, :on_operator_message_edit, :on_operator_message_delete, :allowed_bot_ids

    # @param telegram_bot_token [String, nil] Telegram bot token for downloading files
    # @param slack_bot_token [String, nil] Slack bot token for downloading files
    # @param discord_bot_token [String, nil] Discord bot token (for future use)
    # @param on_operator_message [Proc] Callback when operator sends a message
    #   Receives: session_id, content, operator_name, source_bridge, attachments, reply_to_bridge_message_id
    # @param on_operator_message_with_ids [Proc] Callback when operator sends a message with bridge message ID
    #   Receives: session_id, content, operator_name, source_bridge, attachments, reply_to_bridge_message_id, bridge_message_id
    # @param on_operator_message_edit [Proc] Callback when operator edits a message
    #   Receives: session_id, bridge_message_id, content, source_bridge, edited_at
    # @param on_operator_message_delete [Proc] Callback when operator deletes a message
    #   Receives: session_id, bridge_message_id, source_bridge, deleted_at
    # @param allowed_bot_ids [Array<String>] Optional allowlist of bot IDs for test messages
    def initialize(telegram_bot_token: nil, slack_bot_token: nil, discord_bot_token: nil, on_operator_message: nil, on_operator_message_with_ids: nil, on_operator_message_edit: nil, on_operator_message_delete: nil, allowed_bot_ids: nil)
      @telegram_bot_token = telegram_bot_token
      @slack_bot_token = slack_bot_token
      @discord_bot_token = discord_bot_token
      @on_operator_message = on_operator_message
      @on_operator_message_with_ids = on_operator_message_with_ids
      @on_operator_message_edit = on_operator_message_edit
      @on_operator_message_delete = on_operator_message_delete
      @allowed_bot_ids = allowed_bot_ids || []
    end
  end

  # Handles incoming webhooks from bridges (Telegram, Slack, Discord)
  #
  # @example Usage with Rails
  #   class WebhooksController < ApplicationController
  #     def telegram
  #       handler = WebhookHandler.new(WebhookConfig.new(
  #         telegram_bot_token: ENV['TELEGRAM_BOT_TOKEN'],
  #         on_operator_message: ->(session_id, content, operator_name, source_bridge, attachments, reply_to_bridge_message_id) {
  #           # Handle the message
  #         }
  #       ))
  #       response = handler.handle_telegram_webhook(params.to_unsafe_h)
  #       render json: response
  #     end
  #   end
  class WebhookHandler
    def initialize(config)
      @config = config
    end

    # ─────────────────────────────────────────────────────────────────
    # Telegram Webhook
    # ─────────────────────────────────────────────────────────────────

    # Handle an incoming Telegram webhook
    #
    # @param payload [Hash] The parsed JSON payload from Telegram
    # @return [Hash] Response to send back
    def handle_telegram_webhook(payload)
      return { error: "Telegram not configured" } unless @config.telegram_bot_token

      edited_message = payload["edited_message"]
      if edited_message
        text = edited_message["text"] || ""
        caption = edited_message["caption"] || ""

        return { ok: true } if text.start_with?("/")

        text = caption if text.empty?
        return { ok: true } if text.empty?

        topic_id = edited_message["message_thread_id"]
        return { ok: true } unless topic_id

        if @config.on_operator_message_edit
          bridge_message_id = edited_message["message_id"].to_s
          if bridge_message_id != ""
            @config.on_operator_message_edit.call(
              topic_id.to_s,
              bridge_message_id,
              text,
              "telegram",
              Time.now
            )
          end
        end

        return { ok: true }
      end

      message = payload["message"]
      return { ok: true } unless message

      text = message["text"] || ""
      caption = message["caption"] || ""

      # Skip commands
      return { ok: true } if text.start_with?("/")

      # Use caption if no text
      text = caption if text.empty?

      # Parse media
      media = parse_telegram_media(message)

      # Skip if no content
      return { ok: true } if text.empty? && media.nil?

      # Get topic ID (session identifier)
      topic_id = message["message_thread_id"]
      return { ok: true } unless topic_id

      # Get operator name
      operator_name = message.dig("from", "first_name") || "Operator"

      # Get reply_to_message ID if present (for visual reply linking)
      reply_to_bridge_message_id = message.dig("reply_to_message", "message_id")

      # Download media if present
      attachments = []
      if media
        data = download_telegram_file(media[:file_id])
        if data
          attachments << OperatorAttachment.new(
            filename: media[:filename],
            mime_type: media[:mime_type],
            size: media[:size],
            data: data,
            bridge_file_id: media[:file_id]
          )
        end
      end

      # Call callback
      if @config.on_operator_message
        @config.on_operator_message.call(
          topic_id.to_s,
          text,
          operator_name,
          "telegram",
          attachments,
          reply_to_bridge_message_id
        )
      end
      if @config.on_operator_message_with_ids
        bridge_message_id = message["message_id"].to_s
        if bridge_message_id != ""
          @config.on_operator_message_with_ids.call(
            topic_id.to_s,
            text,
            operator_name,
            "telegram",
            attachments,
            reply_to_bridge_message_id,
            bridge_message_id
          )
        end
      end

      { ok: true }
    end

    # ─────────────────────────────────────────────────────────────────
    # Slack Webhook
    # ─────────────────────────────────────────────────────────────────

    # Handle an incoming Slack webhook
    #
    # @param payload [Hash] The parsed JSON payload from Slack
    # @return [Hash] Response to send back
    def handle_slack_webhook(payload)
      return { error: "Slack not configured" } unless @config.slack_bot_token

      # Handle URL verification challenge
      if payload["type"] == "url_verification" && payload["challenge"]
        return { challenge: payload["challenge"] }
      end

      # Handle event callbacks
      if payload["type"] == "event_callback" && payload["event"]
        event = payload["event"]
        allowed_bot_ids = @config.allowed_bot_ids || []

        return { ok: true } unless event["type"] == "message"

        subtype = event["subtype"]
        if subtype == "message_changed"
          if @config.on_operator_message_edit
            message = event["message"] || {}
            previous = event["previous_message"] || {}
            bot_id = message["bot_id"] || previous["bot_id"] || event["bot_id"]
            if bot_id && !allowed_bot_ids.include?(bot_id)
              return { ok: true }
            end

            thread_ts = message["thread_ts"] || previous["thread_ts"]
            message_ts = message["ts"] || previous["ts"]
            text = message["text"] || ""

            if thread_ts && message_ts
              @config.on_operator_message_edit.call(
                thread_ts,
                message_ts,
                text,
                "slack",
                Time.now
              )
            end
          end

          return { ok: true }
        end

        if subtype == "message_deleted"
          if @config.on_operator_message_delete
            previous = event["previous_message"] || {}
            bot_id = previous["bot_id"] || event["bot_id"]
            if bot_id && !allowed_bot_ids.include?(bot_id)
              return { ok: true }
            end

            thread_ts = previous["thread_ts"]
            message_ts = event["deleted_ts"] || previous["ts"]

            if thread_ts && message_ts
              @config.on_operator_message_delete.call(
                thread_ts,
                message_ts,
                "slack",
                Time.now
              )
            end
          end

          return { ok: true }
        end

        has_content = event["type"] == "message" &&
                      event["thread_ts"] &&
                      (!event["bot_id"] || allowed_bot_ids.include?(event["bot_id"])) &&
                      !event["subtype"]

        files = event["files"] || []
        has_files = files.any?

        if has_content && (event["text"].to_s != "" || has_files)
          thread_ts = event["thread_ts"]
          text = event["text"] || ""

          # Download files if present
          attachments = []
          if has_files
            files.each do |file|
              data = download_slack_file(file)
              if data
                attachments << OperatorAttachment.new(
                  filename: file["name"] || "file",
                  mime_type: file["mimetype"] || "application/octet-stream",
                  size: file["size"] || 0,
                  data: data,
                  bridge_file_id: file["id"]
                )
              end
            end
          end

          # Get operator name
          operator_name = "Operator"
          user_id = event["user"]
          if user_id
            name = get_slack_user_name(user_id)
            operator_name = name if name
          end

          # Call callback (Slack reply support TODO)
          if @config.on_operator_message
            @config.on_operator_message.call(
              thread_ts,
              text,
              operator_name,
              "slack",
              attachments,
              nil
            )
          end
          if @config.on_operator_message_with_ids
            message_ts = event["ts"]
            if message_ts
              @config.on_operator_message_with_ids.call(
                thread_ts,
                text,
                operator_name,
                "slack",
                attachments,
                nil,
                message_ts.to_s
              )
            end
          end
        end
      end

      { ok: true }
    end

    # ─────────────────────────────────────────────────────────────────
    # Discord Webhook
    # ─────────────────────────────────────────────────────────────────

    # Handle an incoming Discord webhook (interactions endpoint)
    #
    # @param payload [Hash] The parsed JSON payload from Discord
    # @return [Hash] Response to send back
    def handle_discord_webhook(payload)
      ping = 1
      application_command = 2
      pong = 1
      channel_message = 4

      interaction_type = payload["type"] || 0

      # Handle PING (verification)
      return { type: pong } if interaction_type == ping

      # Handle Application Commands (slash commands)
      if interaction_type == application_command && payload["data"]
        data = payload["data"]
        if data["name"] == "reply"
          thread_id = payload["channel_id"]
          content = nil

          (data["options"] || []).each do |opt|
            if opt["name"] == "message"
              content = opt["value"]
              break
            end
          end

          if thread_id && content
            # Get operator name
            operator_name = payload.dig("member", "user", "username") ||
                            payload.dig("user", "username") ||
                            "Operator"

            # Call callback (Discord reply support TODO)
            if @config.on_operator_message
              @config.on_operator_message.call(
                thread_id,
                content,
                operator_name,
                "discord",
                [],
                nil
              )
            end

            return {
              type: channel_message,
              data: { content: "✅ Message sent to visitor" }
            }
          end
        end
      end

      { type: pong }
    end

    private

    # ─────────────────────────────────────────────────────────────────
    # Telegram Helpers
    # ─────────────────────────────────────────────────────────────────

    def parse_telegram_media(message)
      if message["photo"].is_a?(Array) && message["photo"].any?
        largest = message["photo"].last
        {
          file_id: largest["file_id"],
          filename: "photo_#{Time.now.to_i}.jpg",
          mime_type: "image/jpeg",
          size: largest["file_size"] || 0
        }
      elsif message["document"]
        doc = message["document"]
        {
          file_id: doc["file_id"],
          filename: doc["file_name"] || "document_#{Time.now.to_i}",
          mime_type: doc["mime_type"] || "application/octet-stream",
          size: doc["file_size"] || 0
        }
      elsif message["audio"]
        audio = message["audio"]
        {
          file_id: audio["file_id"],
          filename: audio["file_name"] || "audio_#{Time.now.to_i}.mp3",
          mime_type: audio["mime_type"] || "audio/mpeg",
          size: audio["file_size"] || 0
        }
      elsif message["video"]
        video = message["video"]
        {
          file_id: video["file_id"],
          filename: video["file_name"] || "video_#{Time.now.to_i}.mp4",
          mime_type: video["mime_type"] || "video/mp4",
          size: video["file_size"] || 0
        }
      elsif message["voice"]
        voice = message["voice"]
        {
          file_id: voice["file_id"],
          filename: "voice_#{Time.now.to_i}.ogg",
          mime_type: voice["mime_type"] || "audio/ogg",
          size: voice["file_size"] || 0
        }
      end
    end

    def download_telegram_file(file_id)
      bot_token = @config.telegram_bot_token
      return nil unless bot_token

      # Get file path
      uri = URI("https://api.telegram.org/bot#{bot_token}/getFile?file_id=#{URI.encode_www_form_component(file_id)}")
      response = Net::HTTP.get_response(uri)
      return nil unless response.is_a?(Net::HTTPSuccess)

      result = JSON.parse(response.body)
      return nil unless result["ok"] && result.dig("result", "file_path")

      file_path = result["result"]["file_path"]

      # Download file
      download_uri = URI("https://api.telegram.org/file/bot#{bot_token}/#{file_path}")
      file_response = Net::HTTP.get_response(download_uri)
      return nil unless file_response.is_a?(Net::HTTPSuccess)

      file_response.body
    rescue StandardError => e
      warn "[WebhookHandler] Telegram file download error: #{e.message}"
      nil
    end

    # ─────────────────────────────────────────────────────────────────
    # Slack Helpers
    # ─────────────────────────────────────────────────────────────────

    def download_slack_file(file)
      download_url = file["url_private_download"] || file["url_private"]
      return nil unless download_url

      uri = URI(download_url)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true

      request = Net::HTTP::Get.new(uri)
      request["Authorization"] = "Bearer #{@config.slack_bot_token}"

      response = http.request(request)
      return nil unless response.is_a?(Net::HTTPSuccess)

      response.body
    rescue StandardError => e
      warn "[WebhookHandler] Slack file download error: #{e.message}"
      nil
    end

    def get_slack_user_name(user_id)
      uri = URI("https://slack.com/api/users.info?user=#{URI.encode_www_form_component(user_id)}")
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true

      request = Net::HTTP::Get.new(uri)
      request["Authorization"] = "Bearer #{@config.slack_bot_token}"

      response = http.request(request)
      return nil unless response.is_a?(Net::HTTPSuccess)

      result = JSON.parse(response.body)
      return nil unless result["ok"]

      result.dig("user", "real_name") || result.dig("user", "name")
    rescue StandardError
      nil
    end
  end
end
